### PR TITLE
feat: day display variant

### PR DESCRIPTION
**Pull request by TobiTenno**
the default value for the variant is `simple`, i.e. the original for this example app

image

this should look familiar, as it's based on https://romcal.js.org

### CODE REVIEW

File: react-app/src/constants/calendars.ts
Comment (TobiTenno) [inline react-app/src/constants/calendars.ts:1]: changes in here are so we don't have to await results from here and they're available on start.

this also moved the overall app to esm, given the tsconfig.json changes

**Comment by tukusejssirs** [comment]
@TobiTenno, how do you run it? It fails to build on my end. :thinking:

---

`pnpm start` throws some errors (always using `nvm use` in the parent folder). The build (in `pnpm start`, i.e. until I see something output on the website) takes for about two minutes.

<details>

  <summary><code>pnpm start</code> errors</summary>

```bash
# rm -rf package-lock.json pnpm-lock.yaml node_modules && git checkout yarn.lock && pnpm i && pnpm start

webpack compiled successfully
ERROR in src/components/LocaleMenu.tsx:4:10
TS2614: Module '"romcal"' has no exported member 'Romcal'. Did you mean to use 'import Romcal from "romcal"' instead?
    2 | import { observer } from 'mobx-react';
    3 | import React, { useContext } from 'react';
  > 4 | import { Romcal } from 'romcal';
      |          ^^^^^^
    5 | import { AppContext } from '../AppContext';
    6 |
    7 | const LocaleMenu = observer(() => {

ERROR in src/components/LocaleMenu.tsx:14:77
TS7006: Parameter 'pkg' implicitly has an 'any' type.
    12 |   const { localeId, setLocaleId } = romcalStore;
    13 |
  > 14 |   const allLocales = Romcal.LOCALE_IDS.reduce((acc: Record<string, string>, pkg, index) => {
       |                                                                             ^^^
    15 |     acc[Romcal.LOCALE_VAR_NAMES[index]] = pkg;
    16 |     return acc;
    17 |   }, {});

ERROR in src/components/LocaleMenu.tsx:14:82
TS7006: Parameter 'index' implicitly has an 'any' type.
    12 |   const { localeId, setLocaleId } = romcalStore;
    13 |
  > 14 |   const allLocales = Romcal.LOCALE_IDS.reduce((acc: Record<string, string>, pkg, index) => {
       |                                                                                  ^^^^^
    15 |     acc[Romcal.LOCALE_VAR_NAMES[index]] = pkg;
    16 |     return acc;
    17 |   }, {});

ERROR in src/components/LocaleMenu.tsx:32:15
TS2322: Type 'unknown' is not assignable to type 'ReactNode'.
    30 |           {Object.entries(allLocales).map(([id, isoId]) => (
    31 |             <option key={id} value={id}>
  > 32 |               {isoId}
       |               ^^^^^^^
    33 |             </option>
    34 |           ))}
    35 |         </NativeSelect>

ERROR in src/RomcalApp.test.tsx:8:23
TS2339: Property 'toBeInTheDocument' does not exist on type 'JestMatchers<HTMLElement>'.
     6 |   render(<RomcalApp />);
     7 |   const textElement = screen.getByText(/A JavaScript library/i);
  >  8 |   expect(textElement).toBeInTheDocument();
       |                       ^^^^^^^^^^^^^^^^^
     9 | });
    10 |

ERROR in src/stores/romcal.store.ts:3:10
TS2614: Module '"romcal"' has no exported member 'Romcal'. Did you mean to use 'import Romcal from "romcal"' instead?
    1 | import { addMonths, subMonths } from 'date-fns';
    2 | import { makeAutoObservable, runInAction } from 'mobx';
  > 3 | import { Romcal, BaseLiturgicalDay } from 'romcal';
      |          ^^^^^^
    4 | import { CALENDARS } from '../constants/calendars';
    5 |
    6 | export class RomcalStore {
```

</details>

---

It works with NPM (`npm start`), whoever, there are some warning output in the terminal, but the website loads, albeit after some long time (which might be an expected side effect of this PR). It look about 1.5 minutes to load the website, while the 4-8 seconds were needed to actually load the website after Webpack packed the code.

<details>

  <summary><code>npm start</code> warnings</summary>

```bash
# rm -rf package-lock.json pnpm-lock.yaml node_modules && git checkout yarn.lock && npm i && npm start

Compiled with warnings.

Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/index.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/index.ts'

Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/named-references.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/named-references.ts'

Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/numeric-unicode-map.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/numeric-unicode-map.ts'

Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/surrogate-pairs.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/surrogate-pairs.ts'

Search for the keywords to learn more about each warning.
To ignore, add // eslint-disable-next-line to the line before.

WARNING in ./node_modules/html-entities/lib/index.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/index.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/index.ts'

WARNING in ./node_modules/html-entities/lib/named-references.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/named-references.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/named-references.ts'

WARNING in ./node_modules/html-entities/lib/numeric-unicode-map.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/numeric-unicode-map.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/numeric-unicode-map.ts'

WARNING in ./node_modules/html-entities/lib/surrogate-pairs.js
Module Warning (from ./node_modules/source-map-loader/dist/cjs.js):
Failed to parse source map from '$repo_root/react-app/node_modules/html-entities/src/surrogate-pairs.ts' file: Error: ENOENT: no such file or directory, open '$repo_root/react-app/node_modules/html-entities/src/surrogate-pairs.ts'

webpack compiled with 4 warnings
```

</details>

---

Also, we might want to upgrade the romcal packages to `v3.0.0-dev.60`.

---

Consider committing `package-lock.json`. I am not sure though whether we should commit `package.json`, `yarn.lock` and `pnpm-lock.yaml` at the same time and if we do so, how it would behave. :thinking: Nevertheless, `npm i` updates `yarn.lock` and creates `package-lock.json`. Running `pnpm i` only creates `pnpm-lock.yaml`.

**Comment by TobiTenno** [comment]
@tukusejssirs 

> how do you run it?

same way as in the readme. I've got warnings on build, but i've always had a slow startup on the website part, but it usually only takes 10-15 seconds after initial server start for the dev server to cache all the assets, it's probably related to loading up all the calendars.

> committing `package-lock.json`

I've committed the updated `yarn.lock`, but i didn't want to switch to `package-lock.json` without consensus, likewise on updating to updating the dev packages. I'm hoping to keep changes in digestible chunks, but i'm happy to do some of these extra if they help out.


> `pnpm start` errors

i'm wondering if those are related to switching a good bit of the processing to ESM-native versions, as that will probably change package resolution, which i'm not sure how pnpm handles those (thus sticking w/ npm)

> `npm start` warnings

I'm seeing the same warnings, but i was seeing similar from just running dependency updates, so they look related to dependencies. i'll see if i can narrow down precisely where they come from, but from the research i'd seen, these are just react-app dependencies, so unfortunately part of trying to stay up-to-date on upstream dependencies

**Comment by TobiTenno** [comment]
> npm warnings

Overriding the version of `html-entities` to the latest seems to remedy this issue. I use an override rather than including it as a dependency because this allows us to not include more in our manifest, just to force our dependencies to use the same good version uniformly.

**Comment by tukusejssirs** [comment]
> I've committed the updated yarn.lock, but i didn't want to switch to package-lock.json without consensus, likewise on updating to updating the dev packages.

FYI: As far as I remember it correctly, Etienne wanted to switch to PNPM in the monorepo.

I personally prefer using PNPM, while I test it with NPM too. I use Yarn only when I see its lock file committed or it is hard-coded somewhere (like in an NPM script).

I was thinking if it would cause some issues if we commit all three lock files. :thinking: I have never done that though.

> I'm hoping to keep changes in digestible chunks, but i'm happy to do some of these extra if they help out.

It might be a good idea to move the lock files discussion out of this PR, however, I think we should talk about it.

> Overriding the version of html-entities to the latest seems to remedy this issue. I use an override rather than including it as a dependency because this allows us to not include more in our manifest, just to force our dependencies to use the same good version uniformly.

How do you override it? Do you need to patching something up?

**Comment by TobiTenno** [comment]
@tukusejssirs 

> How do you override it? Do you need to patch something up?

the `overrides` field in the package.json

File: react-app/package.json [hunk 1]
```diff
@@ -8,73 +8,74 @@
...
+    "@romcal/calendar.wales": "^3.0.0-dev.60",
+    "@textea/json-viewer": "^3.4.0",
     "date-fns": "^2.29.3",
     "mobx": "^6.7.0",
     "mobx-react": "^7.6.0",
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
     "react-scripts": "5.0.1",
-    "romcal": "^3.0.0-dev.57",
+    "romcal": "^3.0.0-dev.60",
```
Comment (tukusejssirs) [inline react-app/package.json:78]: I have found the issue why `pnpm build` fails: the caret.

While using `dev` romcal versions, we should pin the exact romcal version, as we might (and will) introduce some breaking changes in the next `dev` version.

**Review by tukusejssirs — CHANGES_REQUESTED** [review]
I have found the main culprit for the `pnpm build` issue: you used `^` in romcal versions: using it, `^3.0.0-dev.60` resolves to `3.0.0` which is incorrect.

However, I still get an error in a test using `pnpm start` (which si strange: why should a test run on app start? :thinking: However, when I close the error overlay on the website, the app seems to work as expected.

```bash
ERROR in src/RomcalApp.test.tsx:8:23
TS2339: Property 'toBeInTheDocument' does not exist on type 'JestMatchers<HTMLElement>'.
     6 |   render(<RomcalApp />);
     7 |   const textElement = screen.getByText(/A JavaScript library/i);
  >  8 |   expect(textElement).toBeInTheDocument();
       |                       ^^^^^^^^^^^^^^^^^
     9 | });
    10 |
```

File: react-app/package.json (hunk 1 above)
Comment (TobiTenno) [reply react-app/package.json:78]: interesting. i just did a replace... this is also an issue with pnpm, as base npm doesn't have this resolution issue

Comment (tukusejssirs) [reply react-app/package.json:78]: Yeah, it seems to be a PNPM issue, probably related to pnpm/pnpm#6804. I presume that `^3.0.0-dev.*` resolves to any version newer than the specified version, including the stable version (which is released after the `dev` versions). For me it makes sense, although it seems to be a deviation from what NPM does.

I have no idea what resolution is used in NPM in this case.

Anyway, I still think that using exact `dev` version is a good idea. After we use a stable version, we can reintroduce the caret.

Comment (TobiTenno) [reply react-app/package.json:78]: i can definitely change it, but that's kind of unrelated to my changes... this seems like something we'd need to talk about w/ generally any other pnpm issues, cause pnpm is not the context of the example app... 

regardless, i'll get a change in in 10-12h once i'm not working.

Comment (TobiTenno) [reply react-app/package.json]: wasn't as free as I thought i was. it's updated now.

**Comment by TobiTenno** [comment]
> I still get the error described in [here](https://github.com/romcal/romcal-examples/pull/2#pullrequestreview-1926713656) when using PNPM (it does not happen with NPM). I have no idea why would a test be run when starting the app. 🤔
> 
> I’d like to debug this, hopefully later this week.

I'll spin up pnpm to see if i can reproduce it as well, but it's not a test, but i'm not sure either. maybe test is a default in pnpm for pre-start? I'm not entirely sure as it could be related to react-scripts, which i was going to promote avoiding, but makes sense in this specific case for the purposes of an example app being as close to CRA (create-react-app) as possible

**Comment by tukusejssirs** [comment]
Well, `src/RomcalApp.test.tsx` seems to be a test and it errors out on `expect(textElement).toBeInTheDocument()`.

> maybe test is a default in pnpm for pre-start?

There are no default `pre*` scripts in PNPM. Actually, without enabling `enable-pre-post-scripts=true` (e.g. in `.npmrc`), no `pre*` scripts are run when using PNPM.

**Comment by TobiTenno** [comment]
yeah, i know about the parallel test files, but i'll be honest, i haven't touched them much yet. I'll dig through that once I finish some of the linting QOL on the main repo

**Comment by TobiTenno** [comment]
looking at the error, looks like `pnpm start` (i still don't know why, this part i'm not terribly worried about) is running tests that default `npm start` doesn't run. That test was failing, i've now fixed said test.

that said, startup with pnpm took _significantly longer_... I wish i had a timer for this for how long it took to get to first render, but i didn't wanna spend the time doing it.

**Comment by TobiTenno** [comment]
@tukusejssirs is there anything else i can fix before you're ok with the example updates?

**Review by tukusejssirs — APPROVED** [review]

**Comment by tukusejssirs** [comment]
I think the long website startup when using PNPM is caused by the fact that all calendar chunks (e.g. `/static/js/vendors-node_modules_pnpm_romcal_calendar_argentina_3_0_0-dev_60_romcal_3_0_0-dev_60_node_mod-204aff.chunk.js`). IMHO the UI should be loaded ASAP, then the data for the default calendar (the GRC) in the default locale (`en`) should be loaded, the rest of the data should be lazily downloaded _after_ the website is fully loaded from the user/viewer perspective.

However, that might be a task for a separate PR. Note that I am not a frontend dev (I know a bit of HTML and CSS, but never used React, but I used a bit Svelte and Vite 3), therefore, I am not sure if or how to carry out what I suggested above.

---

I think you should upgrade romcal deps to their latest version (`3.0.0-dev.67`). This is quite important IMO.

---

This is probably unrelated to this PR, however, the test fails (both using NPM and PNPM):

```bash
 FAIL  src/__tests__/RomcalApp.test.tsx
  ● Test suite failed to run

    Jest encountered an unexpected token

    Jest failed to parse a file. This happens e.g. when your code or its dependencies use non-standard JavaScript syntax, or when Jest is not configured to support such syntax.

    Out of the box Jest supports Babel, which will be used to transform your files into valid JS based on your Babel configuration.

    By default "node_modules" folder is ignored by transformers.

    Here's what you can do:
     • If you are trying to use ECMAScript Modules, see https://jestjs.io/docs/ecmascript-modules for how to enable it.
     • If you are trying to use TypeScript, see https://jestjs.io/docs/getting-started#using-typescript
     • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
     • If you need a custom transformation specify a "transform" option in your config.
     • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.

    You'll find more details and examples of these config options in the docs:
    https://jestjs.io/docs/configuration
    For information about custom transformations, see:
    https://jestjs.io/docs/code-transformation

    Details:

    /run/media/ts/root/home/ts/git/c10n/romcal-examples__TobiTenno/react-app/src/constants/calendars.ts:10
      GeneralRoman: await Promise.resolve().then(() => (0, _interopRequireWildcard2.default)(require('@romcal/calendar.general-roman'))),
                          ^^^^^^^

    SyntaxError: Unexpected identifier 'Promise'

      2 | import { makeAutoObservable, runInAction } from 'mobx';
      3 | import { Romcal, BaseLiturgicalDay } from 'romcal';
    > 4 | import { CALENDARS } from '../constants/calendars';
        | ^
      5 |
      6 | export class RomcalStore {
      7 |   fetchingData: boolean = false;

      at Runtime.createScriptFromCode (node_modules/.pnpm/jest-runtime@27.5.1/node_modules/jest-runtime/build/index.js:1728:14)
      at Object.<anonymous> (src/stores/romcal.store.ts:4:1)
      at Object.<anonymous> (src/stores/index.ts:1:1)
      at Object.<anonymous> (src/AppContext.tsx:2:1)
      at Object.<anonymous> (src/RomcalApp.tsx:6:1)
      at Object.<anonymous> (src/__tests__/RomcalApp.test.tsx:4:1)
      at TestScheduler.scheduleTests (node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/TestScheduler.js:333:13)
      at runJest (node_modules/.pnpm/@jest+core@27.5.1/node_modules/@jest/core/build/runJest.js:404:19)
```

---

In the end, if you consider this PR done, go ahead and merge it.

**Comment by tukusejssirs** [comment]
## TODO

- [ ] first try it out with these versions:

```bash
pnpm i @mikro-orm/core@6.1.11 @mikro-orm/migrations@6.1.11 @mikro-orm/nestjs@5.2.3 @mikro-orm/postgresql@6.1.11 @mikro-orm/reflection@6.1.11 @mikro-orm/seeder@6.1.11 @mikro-orm/sql-highlighter@1.0.1
pnpm -D i @mikro-orm/cli@6.1.11 typescript@5.1.4
```

- [ ] then try upgrading to their latest versions:

```bash
pnpm i @mikro-orm/core@latest @mikro-orm/migrations@latest @mikro-orm/nestjs@latest @mikro-orm/postgresql@latest @mikro-orm/reflection@latest @mikro-orm/seeder@latest @mikro-orm/sql-highlighter@latest
pnpm -D i @mikro-orm/cli@latest typescript@latest
```

## Create from scratch

```bash
# Install dependencies
pnpm i @mikro-orm/core@6.1.11 @mikro-orm/migrations@6.1.11 @mikro-orm/nestjs@5.2.3 @mikro-orm/postgresql@6.1.11 @mikro-orm/reflection@6.1.11 @mikro-orm/seeder@6.1.11 @mikro-orm/sql-highlighter@1.0.1
pnpm -D i @mikro-orm/cli@6.1.11

# Create a Docker Compose file
cat << EOF > compose.yml
---
name: mikro-orm-migrations
services:
  db:
    env_file:
      - $BASE_PATH/customer/config/postgres/env.conf
    environment:
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: some_long_password
    healthcheck:
      retries: 3
      test: pg_isready -U postgres
      timeout: 5s
    image: timescale/timescaledb-ha:pg16.2-ts2.14.2
    logging:
      options:
        max-file: '7'
        max-size: '20m'
    network_mode: host
    ports:
      - '31234:5321'
    restart: unless-stopped
    shm_size: 256MB
...
EOF

# TODO: Create MikroORM config file
cat << EOF > src/mikro-orm.constant.ts
import {MikroOrmModuleOptions} from '@mikro-orm/nestjs'
import {defineConfig} from '@mikro-orm/postgresql'
import {TsMorphMetadataProvider} from '@mikro-orm/reflection'
import {SqlHighlighter} from '@mikro-orm/sql-highlighter'

export const ORM_CONFIG: MikroOrmModuleOptions = defineConfig({
	allowGlobalContext: true,
	dbName: 'mikro_orm_migrations',
	debug: true,
	discovery: {
		disableDynamicFileAccess: true
	},
	entities: ['./dist/src/entity/*.entity.js'],
	entitiesTs: ['./src/entity/*.entity.ts'],
	highlighter: new SqlHighlighter(),
	host: 'localhost',
	metadataProvider: TsMorphMetadataProvider,
	migrations: {
		allOrNothing: true,
		disableForeignKeys: false,
		dropTables: false,
		emit: 'ts',
		glob: '!(*.d).{js,ts}',
		path: './dist/src/migration',
		pathTs: './src/migration',
		safe: true,
		tableName: 'mikro_orm_migrations',
		transactional: true
	},
	password: 'some_long_password',
	pool: {
		max: 200,
		min: 0
	},
	port: 31234,
	timezone: 'Etc/UTC',
	user: 'user'
})

export default ORM_CONFIG
EOF

# Create folders for entities and migrations
mkdir -p src/{entity,migration}

# TODO: Create a TypeScript file with migrations

# TODO: Add some NPM scripts to `package.json`
# - `docker:up`;
# - `docker:down` (with database data deletion);
# - `start`: start the app;
```



```bash
split --numeric-suffixes 1 -a 1 -l 424 --additional-suffix .txt viv_android_urls_backup.txt viv_android_
split -a 1 -l 424 --additional-suffix .txt viv_android_urls_backup.txt viv_android_
```


---

> tests & jest is a bad tool for the use case

Yeah, Jest is for unit tests (and we should use it for that), however, for UI tests we might want to use Playwright (in a separate PR, OFC).

> new romcal versions

Okay, then do it in a separate PR please. There some imporant bug fixes (#477, #480), however, it is not important from POV of this app.

> the startup time is longer for first launch with my current (awaiting them all in CALENDARS), but it's still long on the startup without the current state.

Subjectively, I feel it is a bit faster than it was before this PR. :man_shrugging:

> i'll see if i can move some of the initial startup waits off of the main thread entirely for getting react to turn stuff on

If you want, you can do this in a separate PR. IMHO the order of loading priority is as follows:

1. UI;
2. the default calendar;
3. the default locale;
4. all the rest.

**Comment by tukusejssirs** [comment]
Sorry for the previous comment; I have published more than intended. I have updated the comment to contain only what is related to this PR.

**Comment by TobiTenno** [comment]
re: speed

yeah, the problem, from what i'm seeing, is that the async imports make it pretty slow. i'm going to try to speed it up a bit, cause i'm not sure why it needs to be awaited at all, so i'm working to optimize the speed a good bit more.

---

i'd definitely like to add better ui testing, because the implementation of how jest works for UI testing is just.... _bad_. 

I'll probably do cypress, but i'll happily take a look at playwright.

---


just to summarize,

- in the current PR:
  - fix load speed related to current changes
- in upcoming prs:
	- update romcal version for all examples
	- add UI testing to replace jest "unit" testing, because jest doesn't work anymore for what we need

if there's anything else, i'd like to make a small set of issues for known stuff to do and i'll link them to this one

**Comment by tukusejssirs** [comment]
> yeah, the problem, from what i'm seeing, is that the async imports make it pretty slow. i'm going to try to speed it up a bit, cause i'm not sure why it needs to be awaited at all, so i'm working to optimize the speed a good bit more.

:+1: :pray:

> i'd definitely like to add better ui testing, because the implementation of how jest works for UI testing is just.... bad.

Yeah, it is not the right tool for the task.

> I'll probably do cypress, but i'll happily take a look at playwright.

Use whatever you want. Either is fine with me, I have just read that Playwright might be better/preferred these days, however, I am not a frontend dev. :man_shrugging:

> just to summarize,
>
> - in the current PR:
>   - fix load speed related to current changes
> - in upcoming prs:
> 	- update romcal version for all examples (#3)
> 	- add UI testing to replace jest "unit" testing, because jest doesn't work anymore for what we need (#4)
>
> if there's anything else, i'd like to make a small set of issues for known stuff to do and i'll link them to this one

I agree with this plan, however, I do think that for _actual_ unit tests (e.g. functions, methods), Jest does a good job, however, not for UI tests.

Thanks! :pray:

**Pull request merged**